### PR TITLE
NFT (natural feature tracking) on AR.js

### PR DESCRIPTION
**Pull request by kalwalt**
**⚠️ All PRs have to be done versus 'dev' branch, so be aware of that, or we'll close your issue ⚠️**

**What kind of change does this PR introduce?**
this PR want to test the new jsartoolkit5 from https://github.com/kalwalt/jsartoolkit5/pull/1 with the new changes especially the NFT part.

### **🖼 This will eventually make AR.js capable of detect and track 2D images.**

https://twitter.com/nicolocarp/status/1204841439160684549

Could be raise some bugs and incompatibilties between the code. It is more a test...
**Can it be referenced to an Issue? If so what is the issue # ?**

From jsartoolkit5 repo:  PR  https://github.com/kalwalt/jsartoolkit5/pull/1
From AR.js issues:

* https://github.com/jeromeetienne/AR.js/issues/544
* https://github.com/jeromeetienne/AR.js/issues/149
* https://github.com/jeromeetienne/AR.js/issues/503
* https://github.com/jeromeetienne/AR.js/issues/420
* https://github.com/jeromeetienne/AR.js/issues/227
* https://github.com/jeromeetienne/AR.js/issues/190
* https://github.com/jeromeetienne/AR.js/issues/185

also, a repo cleaning will help for: https://github.com/jeromeetienne/AR.js/issues/433

**How can we test it?**
Just test the examples and tell us what happens...

**Summary**

**Does this PR introduce a breaking change?**
It could be... but it will add NFT feature.

**Please TEST your PR before proposing it. Specify here what device you have used for tests, version of OS and version of Browser**

**Other information**
I uploaded the new jsartookit5 libs and build the AR.js libs with makefile. 

**Progress**
- [x] Clean repo before starting the porting
- [x] Import new jsartoolkit5 version, coming with NFT
- [x] Adapt the code (threejs side) to support NFT (emulate what's already done for fiducial markers)
- [ ] Fix the rendering of content on AR.js side, for NFT (onMarkerFound function)
- [ ] Remove all rendering code on the Worker
- [ ] Adapt the code (aframe side) to support NFT (emulate what's already done for fiducial markers)
- [ ] Create at least one example for threejs + NFT
- [ ] Create at least one example for aframe + NFT
- [ ] Change documentation (principal Readme + Readmes inside threejs + aframe folders)
- [ ] release version 3 of AR.js

### CODE REVIEW

**Comment by kalwalt** [comment]
The simplest examples don't works i receive an error like:

```js
artoolkit.api.js:1235 Uncaught TypeError: artoolkit.setup is not a function
    at ARController._initialize (artoolkit.api.js:1235)
    at new ARController (artoolkit.api.js:128)
    at ARCameraParam.onload (threex-artoolkitcontext.js:197)
    at ARCameraParam.<anonymous> (artoolkit.api.js:1673)
    at writeCallback (artoolkit.api.js:1892)
    at writeByteArrayToFS (artoolkit.api.js:1919)
    at XMLHttpRequest.oReq.onload (artoolkit.api.js:1936)
ARController._initialize @ artoolkit.api.js:1235
ARController @ artoolkit.api.js:128
(anonymous) @ threex-artoolkitcontext.js:197
(anonymous) @ artoolkit.api.js:1673
writeCallback @ artoolkit.api.js:1892
writeByteArrayToFS @ artoolkit.api.js:1919
oReq.onload @ artoolkit.api.js:1936
load (async)
ajax @ artoolkit.api.js:1931
loadCamera @ artoolkit.api.js:1900
ARCameraParam.load @ artoolkit.api.js:1670
ARCameraParam @ artoolkit.api.js:1649
ARjs.Context._initArtoolkit @ threex-artoolkitcontext.js:195
ARjs.Context.init @ threex-artoolkitcontext.js:106
(anonymous) @ dev.html:114
```
you can try different examples ( basic, default, dev..) but same result. we have to change the code but i have not a clear idea if jsartoolkit5 or AR.js or both...

**Comment by kalwalt** [comment]
with the latest commit small example to test, i receive this error:

```shell
THREE.WebGLRenderer 103
test-new.html:86 Live reload enabled.
773test-new.html:53 Uncaught ReferenceError: arToolkitContext is not defined
    at onResize (test-new.html:53)
    at onReady (test-new.html:43)
    at onSourceReady (ar.js:6275)
    at ar.js:6420
onResize @ test-new.html:53
onReady @ test-new.html:43
onSourceReady @ ar.js:6275
(anonymous) @ ar.js:6420
setInterval (async)
success @ ar.js:6418
Promise.then (async)
(anonymous) @ ar.js:6404
Promise.then (async)
ARjs.Source._initSourceWebcam @ ar.js:6379
ARjs.Source.init @ ar.js:6251
(anonymous) @ test-new.html:42
```

**Comment by kalwalt** [comment]
It's enough for today! i will continue another time...

**Comment by kalwalt** [comment]
i changed the **test-new** example a bit and i'm testing a simple example with NFT. Basically like the video_threejs_worker.html example in the nft_improved_worker folder using directly the artoolkit functions... but i get an error with the camera initailization, see the log:

```js
Live reload enabled.
three.js:22661 THREE.WebGLRenderer 103
three.js:47439 THREE.MeshNormalMaterial: .shading has been removed. Use the boolean .flatShading instead.
set @ three.js:47439
start @ threejs_worker.js:59
document.getElementById.onclick @ video_threejs_worker.html:49
video_threejs_worker.html:48 [Violation] 'click' handler took 165ms
artoolkit.min.js:1 onload callback should be defined
ARCameraParam @ artoolkit.min.js:1
load @ worker.js:24
self.onmessage @ worker.js:7
artoolkit.min.js:1 onerror callback should be defined
ARCameraParam @ artoolkit.min.js:1
load @ worker.js:24
self.onmessage @ worker.js:7
artoolkit.min.js:1 [warning] *** Camera Parameter resized from 640, 480. ***
put_char @ artoolkit.min.js:1
write @ artoolkit.min.js:1
write @ artoolkit.min.js:1
doWritev @ artoolkit.min.js:1
_fd_write @ artoolkit.min.js:1
___wasi_fd_write @ artoolkit.min.js:1
vx @ artoolkit.min.js:13
_x @ artoolkit.min.js:13
Tx @ artoolkit.min.js:13
Sx @ artoolkit.min.js:13
_e @ artoolkit.min.js:10
Ze @ artoolkit.min.js:10
gx @ artoolkit.min.js:13
qu @ artoolkit.min.js:7
ex @ artoolkit.min.js:13
Y_ @ artoolkit.min.js:14
dynCall_iiiii_10 @ VM16:2
setup @ VM61:8
ARController._initialize @ artoolkit.min.js:1
ARController @ artoolkit.min.js:1
param.onload @ worker.js:26
(anonymous) @ artoolkit.min.js:1
writeCallback @ artoolkit.min.js:1
writeByteArrayToFS @ artoolkit.min.js:1
oReq.onload @ artoolkit.min.js:1
load (async)
ajax @ artoolkit.min.js:1
loadCamera @ artoolkit.min.js:1
ARCameraParam.load @ artoolkit.min.js:1
ARCameraParam @ artoolkit.min.js:1
load @ worker.js:24
self.onmessage @ worker.js:7
artoolkit.min.js:1 [info] Allocated videoFrameSize 307200
put_char @ artoolkit.min.js:1
write @ artoolkit.min.js:1
write @ artoolkit.min.js:1
doWritev @ artoolkit.min.js:1
_fd_write @ artoolkit.min.js:1
___wasi_fd_write @ artoolkit.min.js:1
vx @ artoolkit.min.js:13
_x @ artoolkit.min.js:13
Tx @ artoolkit.min.js:13
Sx @ artoolkit.min.js:13
_e @ artoolkit.min.js:10
Ze @ artoolkit.min.js:10
qu @ artoolkit.min.js:7
ex @ artoolkit.min.js:13
Y_ @ artoolkit.min.js:14
dynCall_iiiii_10 @ VM16:2
setup @ VM61:8
ARController._initialize @ artoolkit.min.js:1
ARController @ artoolkit.min.js:1
param.onload @ worker.js:26
(anonymous) @ artoolkit.min.js:1
writeCallback @ artoolkit.min.js:1
writeByteArrayToFS @ artoolkit.min.js:1
oReq.onload @ artoolkit.min.js:1
load (async)
ajax @ artoolkit.min.js:1
loadCamera @ artoolkit.min.js:1
ARCameraParam.load @ artoolkit.min.js:1
ARCameraParam @ artoolkit.min.js:1
load @ worker.js:24
self.onmessage @ worker.js:7
artoolkit.min.js:1 Uncaught TypeError: errorCallback is not a function
    at XMLHttpRequest.oReq.onload (artoolkit.min.js:1)
oReq.onload @ artoolkit.min.js:1
load (async)
ajax @ artoolkit.min.js:1
addNFTMarker @ artoolkit.min.js:1
ARController.loadNFTMarker @ artoolkit.min.js:1
param.onload @ worker.js:33
(anonymous) @ artoolkit.min.js:1
writeCallback @ artoolkit.min.js:1
writeByteArrayToFS @ artoolkit.min.js:1
oReq.onload @ artoolkit.min.js:1
load (async)
ajax @ artoolkit.min.js:1
loadCamera @ artoolkit.min.js:1
ARCameraParam.load @ artoolkit.min.js:1
ARCameraParam @ artoolkit.min.js:1
load @ worker.js:24
self.onmessage @ worker.js:7
```
but maybe the problem is not here... i will see.

**Comment by kalwalt** [comment]
Ok now The NFT example works but without modifications to THREEX or other part of the code in AR.js, this is another problem to solve. First we need to integrate the new artoolkit (with NFT) into AR.js.

**Comment by nicolocarpignoli** [comment]
We have to see how AR.js integrates with artoolkit event handling. When marker are detected, for example (but also other stuff) callbacks are called or events are handled, from artoolkit to AR.js.
We have to replicate exactly what is done for fiducial markers, this time for NFT.

Technically it is not difficult, theorically speaking. But there will be a lot of problem solving :)

**Comment by kalwalt** [comment]
> We have to see how AR.js integrates with artoolkit markers handling. When marker are detected, for example (but also other stuff) callbacks are called or events are handled, from artoolkit to AR.js.
> We have to replicate exactly what is done for fiducial markers, this time for NFT.

Yes, i understand this. it needs to understand step by step. I will do a summary and some considerations above the jsartoolkit5 changes, i think this will help us.

> Technically it is not difficult, theorically speaking. But there will be a lot of problem solving :)

Yes probably for you is easier! :smiley: but for me require a lot...

**Comment by kalwalt** [comment]
Thank you @nicolocarpignoli for the latest push and merge!

**Comment by kalwalt** [comment]
running the **threejs/examples/basic.html** example i get:
```js
Uncaught ReferenceError: importScripts is not defined
    at ar.js:2000
(anonymous) @ ar.js:2000

```
because artoolkit.worker.js i suppose.

File: three.js/Makefile [hunk 1]
```diff
@@ -4,7 +4,8 @@ watch: build
 build-main:
 	echo	> build/ar.js
 	cat 	vendor/jsartoolkit5/build/artoolkit.min.js	\
-		vendor/jsartoolkit5/js/artoolkit.api.js		>> build/ar.js
+		vendor/jsartoolkit5/js/artoolkit.api.js		\
```
Comment (kalwalt) [inline three.js/Makefile]: maybe artoolkit.api.js is not needed anymore... will test this

File: three.js/Makefile [hunk 2]
```diff
@@ -17,7 +18,8 @@ minify-main: build-main
 build-lean:
 	echo > build/ar.lean.js
 	cat vendor/jsartoolkit5/build/artoolkit.min.js \
-		vendor/jsartoolkit5/js/artoolkit.api.js >> build/ar.lean.js
+		vendor/jsartoolkit5/js/artoolkit.api.js \
```
Comment (kalwalt) [inline three.js/Makefile]: same here... as the previuos comment

File: three.js/Makefile [hunk 3]
```diff
@@ -17,7 +18,8 @@ minify-main: build-main
...
+		vendor/jsartoolkit5/js/artoolkit.worker.js 	>> build/ar.lean.js
```
Comment (kalwalt) [inline three.js/Makefile]: see this [comment](https://github.com/jeromeetienne/AR.js/pull/674#issuecomment-568770206)

**Comment by kalwalt** [comment]
We removed the lean version of the lib because has not sense, there is not a lean version anymore.

**Comment by kalwalt** [comment]
importScripts issue solved thanks to this article https://stackoverflow.com/questions/14500091/uncaught-referenceerror-importscripts-is-not-defined

File: three.js/src/threex/threex-armarkercontrols.js [hunk 4]
```diff
@@ -210,6 +210,10 @@ ARjs.MarkerControls.prototype.name = function(){
 		name += ' - ' + basename
 	}else if( this.parameters.type === 'barcode' ){
 		name += ' - ' + this.parameters.barcodeValue
+	}else if( this.parameters.type === 'nft' ){
+		var url = this.parameters.patternUrl
+		var basename = url.replace(/^.*\//g, '')
```
Comment (kalwalt) [inline three.js/src/threex/threex-armarkercontrols.js]: not sure if it is needed this...

File: three.js/src/threex/threex-armarkercontrols.js [hunk 5]
```diff
@@ -251,6 +255,8 @@ ARjs.MarkerControls.prototype._initArtoolkit = function(){
 		}else if( _this.parameters.type === 'barcode' ){
 			artoolkitMarkerId = _this.parameters.barcodeValue
 			arController.trackBarcodeMarkerId(artoolkitMarkerId, _this.parameters.size);
+		}else if( _this.parameters.type === 'nft' ){
+			console.log('nft test');
```
Comment (kalwalt) [inline three.js/src/threex/threex-armarkercontrols.js]: Here we could import the artoolkit.worker.js into the Worker but it require a path,
example code:

```js
// init the worker
var worker = new Worker ('path/to/artoolkit.worker.js');
```

**Comment by nicolocarpignoli** [comment]
In my opinion we should first reach a stable point on jsartoolkit5 repository with the NFT version and all demos working. Use web assembly as default maybe, if gives better performances and its supported the same as the version only with workers. 

After that we should restart work on this pr to integrate it on arjs and also think of a way to integrate the NFT generator on this repository or maybe on an arjs org

**Comment by kalwalt** [comment]
I tested the latest, yes something is showing very instable at the moment, but for sure could be improved! Bravo Nicolò!

**Comment by kalwalt** [comment]
i think the flashing of the model happens because we should add the `getNFTmarker` event listeners to get data from the NFT marker.

**Comment by kalwalt** [comment]
We could create in jsartoolkit5 (artoolkit.api.js) a NFT_MARKER=2 in the exported api, and add marker type into the NFT dispatch event. This should be help us to retieve the type data.

**Comment by nicolocarpignoli** [comment]
> We could create in jsartoolkit5 (artoolkit.api.js) a NFT_MARKER=2 in the exported api, and add marker type into the NFT dispatch event. This should be help us to retieve the type of data.

yes I saw that was missing! next step is to add that on jsartoolkit5 side in my opinion

**Comment by kalwalt** [comment]
Now it works! :smile:

**Comment by kalwalt** [comment]
@nicolocarpignoli I will do that changes inside jsartoolkit5, if i can this night...

**Comment by kalwalt** [comment]
i did the changes to artoolkit.api.js in this commit https://github.com/kalwalt/jsartoolkit5/pull/1/commits/9066705237c4ef4e699c1dfc86abf458aa7280ef if you updates the jsartoolkit lib make a ref to this commit.

**Comment by kalwalt** [comment]
With https://github.com/jeromeetienne/AR.js/pull/674/commits/7672bf3aaf1aec1cf8a3bf3e84e98c03ec42d67e the **ar.js** libs can be minifyed with uglifyjs using the ['use strict'](url) directive at the beggining of the jsartoolkit5 lib (artoolkit.min.js)

**Comment by kalwalt** [comment]
@nicolocarpignoli i tested your work! Now we have **NFT** + **webWorker** inside Ar.js !! :1st_place_medal: :fire: :tada: 
Thinking if we can improve the code in some way...

**Comment by kalwalt** [comment]
We should also see if it is necessary 

- all the THREE.js code,  

- all the code regarding the canvas, video stilying etc

- improve the loading time experience ( apply the gif loading as in jsarttolkit5 )

- a check if the WebWorker standard is supported by the browser with a console log error if it not supported.

**Comment by nicolocarpignoli** [comment]
> @nicolocarpignoli i tested your work! Now we have **NFT** + **webWorker** inside Ar.js !! 🥇 🔥 🎉
> Thinking if we can improve the code in some way...

of course we can :)

File: three.js/src/threex/threex-arnftworker.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
+var THREEx = THREEx || {}
+
+THREEx.ArNFTWorker = function(object3d){
+  //this.id = id;
+  this.object3d = object3d;
+}
+
+Object.assign( THREEx.ArNFTWorker.prototype, THREE.EventDispatcher.prototype );
+
+THREEx.ArNFTWorker.prototype.isMobile = function()  {
```
[the remark this answers is not in the document]
Comment (nicolocarpignoli) [reply three.js/src/threex/threex-arnftworker.js]: yes good idea, we can remove the isMobile for that one

**Comment by nicolocarpignoli** [comment]
> We should also see if it is necessary
> 
> * all the THREE.js code,
> * all the code regarding the canvas, video stilying etc
> * improve the loading time experience ( apply the gif loading as in jsarttolkit5 )
> * a check if the WebWorker standard is supported by the browser with a console log error if it not supported.

regarding the last one, no problem for the Web Worker support: https://caniuse.com/#feat=webworkers

also for web assembly, it's ok: https://caniuse.com/#feat=wasm (IE not a problem in my opinion)

**Comment by kalwalt** [comment]
> regarding the last one, no problem for the Web Worker support: https://caniuse.com/#feat=webworkers

> also for web assembly, it's ok: https://caniuse.com/#feat=wasm (IE not a problem in my opinion)
yeah probably it is not needed... will see that.

File: three.js/src/threex/threex-arnftworker.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (kalwalt) [reply three.js/src/threex/threex-arnftworker.js]: No this is not possible, but maybe we can move the isMobile() function outside the ArNFTWorker object inside maybe ARjs.Context (threex-artoolkitcontext) ? This si only a detail the the code will work in any case...

**Comment by kalwalt** [comment]
@nicolocarpignoli the code is not working with the worker, you didn't rebuild the libs with the new changes...  i built the libs and i'm getting this error:
```
THREE.WebGLRenderer 103
ar.min.js:1 onerror callback should be defined
three.min.js:931 THREE.MeshNormalMaterial: .shading has been removed. Use the boolean .flatShading instead.
nft.html:186 Live reload enabled.
ar.min.js:1 [info] Allocated videoFrameSize 1228800
ar.min.js:1 [info] Pattern detection mode set to 1.
ar.min.js:1 [info] Pattern ratio size set to 0.500000.
ar.min.js:1 [Violation] 'load' handler took 197ms
three.min.js:184 THREE.WebGLRenderer 103
three.min.js:188 THREE.WebGLRenderer: Error creating WebGL context.
three.min.js:187 Uncaught Error: Error creating WebGL context.
    at new ge (three.min.js:187)
    at THREEx.ArNFTWorker.start (ar.min.js:1)
    at ar.min.js:1
    at onSourceReady (ar.min.js:1)
    at success (ar.min.js:1)
nft.html:147 [Violation] 'requestAnimationFrame' handler took 104ms
nft.html:147 [Violation] 'requestAnimationFrame' handler took 164ms
nft.html:147 [Violation] 'requestAnimationFrame' handler took 58ms
```

**Comment by kalwalt** [comment]
Probably we need to take the **THREE.WebGLRenderer**  from the artoolkit source or when it is initailized inside the html and attach it to the ArNFTWorker object in some way.

**Comment by kalwalt** [comment]
with the latest: 
```
THREE.WebGLRenderer 103
ar.min.js:1 onerror callback should be defined
three.min.js:931 THREE.MeshNormalMaterial: .shading has been removed. Use the boolean .flatShading instead.
nft.html:186 Live reload enabled.
ar.min.js:1 [info] Allocated videoFrameSize 1228800
ar.min.js:1 [info] Pattern detection mode set to 1.
ar.min.js:1 [info] Pattern ratio size set to 0.500000.
ar.min.js:1 [Violation] 'load' handler took 188ms
three.min.js:525 THREE.Object3D.add: object not an instance of THREE.Object3D. undefined
ar.min.js:1 Uncaught ReferenceError: isMobile is not defined
    at load (ar.min.js:1)
    at THREEx.ArNFTWorker.start (ar.min.js:1)
    at ar.min.js:1
    at onSourceReady (ar.min.js:1)
    at success (ar.min.js:1)
nft.html:147 [Violation] 'requestAnimationFrame' handler took 95ms
nft.html:147 [Violation] 'requestAnimationFrame' handler took 51ms
﻿
```
 i think the ArNFTWorker needs some changes.
​

**Comment by nicolocarpignoli** [comment]
`Uncaught ReferenceError: isMobile is not defined`

is the first problem to look at

**Comment by kalwalt** [comment]
3 issue to solve:
- the code fails at this line when attaching the THREE.3DObject:

https://github.com/kalwalt/AR.js/blob/45a45e0f3e66613046923fd066aa0e91c7dcce74/three.js/src/threex/threex-arnftworker.js#L53

- when attaching the renderer:

https://github.com/kalwalt/AR.js/blob/45a45e0f3e66613046923fd066aa0e91c7dcce74/three.js/src/threex/threex-arnftworker.js#L83

- and passing the imageData 

https://github.com/kalwalt/AR.js/blob/45a45e0f3e66613046923fd066aa0e91c7dcce74/three.js/src/threex/threex-arnftworker.js#L172

**Comment by cvasquez-github** [comment]
Hey guys, thank you for working on this! the whole AR ecosystem developers really appreciate this huge effort of yours! This feature is so important that I would love to contribute some way, feel free to let me know if you need a Beta tester or anything!! Thanks again.

**Comment by kalwalt** [comment]
> Hey guys, thank you for working on this! the whole AR ecosystem developers really appreciate this huge effort of yours! This feature is so important that I would love to contribute some way, feel free to let me know if you need a Beta tester or anything!! Thanks again.

@cvasquez-github you are very welcome! if you want to test the code no problem at all! and if you have some suggestion or other type of advise even better... :smile: 

@nicolocarpignoli with the latest commit https://github.com/jeromeetienne/AR.js/pull/674/commits/a39c4ee7f19bc505eaab9645ce4ba2d45ab09e92 finally the Worker is used! But the code load the external **artoolkit.worker.js** library and as a consequence also the external **artoolkit.min.js**, maybe the Blob object may solve this?
In this scenario two ArController and ARCameraParam are created. the ones in te Worker and the ones in the non Worker environment.  We should think about this aspect now as we talked....

**Comment by kalwalt** [comment]
@cvasquez-github you can also test my PR https://github.com/kalwalt/jsartoolkit5/pull/1 for jsartoolkit5 and make some tests on this subject https://github.com/kalwalt/jsartoolkit5/issues/28 this one will be very useful indeed.

**Comment by kalwalt** [comment]
Tested with the pinball image. The marker is tracked but no model is rendered/displayed. Should see why this happens.

**Comment by cvasquez-github** [comment]
Of course! I think might need some help on how to train the image detection for the NFT. Or should I just use the "pinball" image to try it (https://imgur.com/HvcmRVl)?

**Comment by kalwalt** [comment]
@cvasquez-github just use the pinball image (as you posted the link) thanks!

**Comment by nicolocarpignoli** [comment]
@kalwalt updated the checklist: https://github.com/jeromeetienne/AR.js/pull/674#issue-352547279

**Comment by cvasquez-github** [comment]
I will run some tests now. How can I set the NTF marker in aframe-ar project?  I have the 3 files .fset .fset3 and .iset. I tried pointing the a-marker url to the pinball folder like the threejs example but did not work. Thanks!

**Comment by kalwalt** [comment]
@cvasquez-github for now we are only working on the threejs side, aframe integration will be in another moment. Actually in this branch we are breaking things frequently, we did it working without the WebWorker but there was an amount of not needed stuff as Three.js rendering routines that wasn't necessary. Alternatively you can test instead the jsartoolkit5 examples as i said in the above comment.

**Comment by kalwalt** [comment]
We are continuing development for NFT. Right now we are testing a new approach in another branch https://github.com/kalwalt/AR.js/tree/clean-worker. The model (the sphere) appears intermittently unstable. Nicolo and I think it's probably related to how data is passed to the matrix see here 

https://github.com/kalwalt/AR.js/blob/0b5c08b909bac9a93d62606480731626190ede07/three.js/src/threex/threex-armarkercontrols.js#L282 

and
      https://github.com/kalwalt/AR.js/blob/0b5c08b909bac9a93d62606480731626190ede07/three.js/src/threex/threex-armarkercontrols.js#L326

@misdake @hatsumatsu do you think that is correct? Any help will be really appreciated, thanks in advance. :smile:

**Comment by nicolocarpignoli** [comment]
We are creating a new PR starting from this one, but on a different branch (`v3`)

**Pull request closed without merging**